src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineTraceForestBuilder.kt

```diff
@@ -0,0 +1,258 @@
+}
+
+private fun SuspendContextImpl.createCoroutineTraceForest(
+    traces: List<CoroutineTrace?>,
+    lastRunningStackFrame: String
+): JBScrollPane? {
+    if (traces.isEmpty()) return null
+    val vertexData: MutableList<JBList<String>?> = mutableListOf()
+    val componentData: MutableList<Component> = mutableListOf()
+    var previousListSelection: JBList<*>? = null
+    var maxWidth = 0
+    var traceNotNullCount = 0
+
+    traces.forEach { trace ->
+        if (trace == null) {
+            vertexData.add(null)
+            return@forEach
+        }
+
+        val vertex = CoroutineTraceUI(trace, lastRunningStackFrame, this).getVertex()
+        vertex.addListSelectionListener { e ->
+            val currentList = e.source as? JBList<*> ?: return@addListSelectionListener
+            if (previousListSelection != currentList) {
+                previousListSelection?.clearSelection()
+            }
+            previousListSelection = currentList
+        }
+        vertexData.add(vertex)
+        maxWidth += vertex.preferredWidth
+        traceNotNullCount += 1
+    }
+
+    val averagePreferredWidth = maxWidth / traceNotNullCount
+    vertexData.forEach { vertex ->
+        if (vertex != null) {
+            vertex.preferredWidth = averagePreferredWidth
+            componentData.add(vertex)
+            return@forEach
+        }
+        componentData.add(Separator())
+    }
+
+    val forest = ContainerWithEdges()
+    componentData.forEach { forest.add(it) }
+    forest.layout = ForestLayout()
+
+    return JBScrollPane(forest)
+}
+
+private fun createCoroutineTraces(rootValue: Node): List<CoroutineTrace?> {
+    val stack = Stack<Pair<Node, Int>>().apply { push(rootValue to 0) }
+    val parentStack = Stack<Node>()
+    var previousLevel: Int? = null
+    val coroutineTraces = mutableListOf<CoroutineTrace?>()
+
+    while (stack.isNotEmpty()) {
+        val (currentNode, currentLevel) = stack.pop()
+        val parent = if (parentStack.isNotEmpty()) parentStack.pop() else null
+
+        if (parent != null && parent.num != currentNode.num) {
+            val currentTrace = CoroutineTrace(
+                mutableListOf(currentNode.stackFrameItem),
+                CoroutineStacksBundle.message("number.of.coroutines.active", currentNode.num),
+                currentNode.coroutinesActive
+            )
+            repeat((previousLevel ?: 0) - currentLevel + 1) {
+                coroutineTraces.add(null)
+            }
+            coroutineTraces.add(currentTrace)
+            previousLevel = currentLevel
+        } else if (parent != null) {
+            coroutineTraces.lastOrNull()?.stackFrameItems?.add(0, currentNode.stackFrameItem)
+        }
+
+        currentNode.children.values.reversed().forEach { child ->
+            stack.push(child to (currentLevel + 1))
+            parentStack.push(currentNode)
+        }
+    }
+
+    return coroutineTraces
+}
+
+private class CoroutineTraceUI(
```

Comment:
I think it would be better to rewrite this class like:
```
private class CoroutineFramesList(
    suspendContext: SuspendContextImpl,
    trace: CoroutineTrace,
    lastRunningStackFrame: String
) : JBList<String>() {
    init {
        val debugProcess = suspendContext.debugProcess
        val activeExecutionStack = suspendContext.activeExecutionStack

        val data = mutableListOf<String>()
        data.add(trace.header)
        data.addAll(trace.stackFrameItems.map { it.toString() })
        setListData(data.toTypedArray())
        val lastStackFrame = data.getOrNull(1)

        border = if (lastRunningStackFrame == lastStackFrame) {
            createRoundedBorder(JBColor.BLUE)
        } else {
            createRoundedBorder(JBColor.BLACK)
        }

        cellRenderer = CoroutineStacksPanel.CustomCellRenderer(trace.coroutinesActiveLabel)
        addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent?) {
                val list = e?.source as? JBList<*> ?: return
                val index = list.locationToIndex(e.point).takeIf { it > 0 } ?: return
                val stackFrameItem = trace.stackFrameItems[index - 1]

                stackFrameItem?.let { frameItem ->
                    val frame = frameItem.createFrame(debugProcess)

                    if (activeExecutionStack != null && frame != null) {
                        suspendContext.setCurrentStackFrame(activeExecutionStack, frame)
                    }
                }
            }
        })
    }
}
```

Reply:
And also let's move it to a separate file! Together with `createRoundedBorder` and  `CustomCellRenderer`